### ISSUE TITLE
`shared_library` support

### ISSUE THREAD

**Issue by indutny** [indutny/gyp.js#2]
gyp.js currently supports `static_library`, `executable`, and `none` `target_type`, but `shared_library` is very important for Node.js addons! Therefore we need to support it in `ninja`.

Hints:

* New ninja rule `solink` needs to be added [here][0]
* This rule should be used [there][1]

Possible problems - many.

Support:

- [ ] OS X
- [ ] Windows
- [ ] Linux
- [ ] FreeBSD
- [ ] Others

[0]: https://github.com/indutny/gyp.js/blob/710209317b573119c75f8e2d3f62802da5075583/lib/gyp/generator/ninja/index.js#L494-L503
[1]: https://github.com/indutny/gyp.js/blob/710209317b573119c75f8e2d3f62802da5075583/lib/gyp/generator/ninja/index.js#L366-L370

**Comment by indutny** [comment]
Could be tested on https://github.com/indutny/llnode

**Comment by indutny** [comment]
I'm going to take OS X support for this.

**Comment by indutny** [comment]
OS X - done.

**Comment by indutny** [comment]
Linux - appears to be working.

**Comment by indutny** [comment]
I'm inclined to say that FreeBSD and other Unixes should work, but this needs a verification.

**Comment by indutny** [comment]
Anyone wants to verify this?

**Comment by kapouer** [comment]
Hey i'd like to ! Could you tell me where i should start ? llnode seems to be something different.

**Comment by indutny** [comment]
@kapouer thank you!

@pmed you said that you tested windows shared_library support on node.js addons, would you care to share some info on this with us? Thanks!

**Comment by indutny** [comment]
@kapouer `llnode` may be a bit too complicated. However, if you want to give it a shot, the way to do it is:

* Install node.js on the system
* Install `ninja` from https://ninja-build.org/
* `npm install -g gyp.js`
* `git clone git@github.com:indutny/llnode.git && cd llnode`
* Install lldb-3.8 or any lldb, and its headers (Honestly, I don't know how to do it on FreeBSD, or anywhere else except Mac and Linux)
* run `gyp -Icommon.gypi llnode.gyp`
* run `ninja -C out/Release`
* `ls -la out/Release`
* should have `llnode.so`

**Comment by indutny** [comment]
Alright, I have pushed `gyppies` to the repo. So it should be possible to test this by just running `npm test`. Note that tests should be run from MSVS console on Windows, and the `NINJA` env variable should be set to the full path to `ninja.exe`.

**Comment by pmed** [comment]
@indutny, I've forked `node-gyp` to use `gyp.js` instead of `gyp`: https://github.com/pmed/node-gyp/tree/gyp.js

To build a native addon you just need a C++ compiler and Ninja installed. Visual C++ is used by default on Windows (my version of `node-gyp` tries to locate a VC++ version installed in the system).

And ninja executable should be in a `PATH`. Yoг can also set `NINJA` environment variable or set `-ninja` command line option for `node-gyp` to override the ninja location.

ps. I just saw you added `platform.win.genEnvironment()` that performs similar to [findMSVS()](https://github.com/pmed/node-gyp/blob/gyp.js/lib/build.js#L77) function. Probably I would try to use your implementation.

**Comment by indutny** [comment]
@pmed Fantastic!

**Comment by indutny** [comment]
@pmed I have just pushed some fixes, and it appears that all `npm test`s pass on Windows, except the `copy` one.

**Comment by indutny** [comment]
@pmed One more thought: what do you think about contributing your patch back to node-gyp, with `gyp.js` used under a command-line flag or env variable?

**Comment by indutny** [comment]
See: https://github.com/nodejs/node-gyp/issues/960

**Comment by pmed** [comment]
Yes, command-line flag to switch between gyp and gyp.js  would be great. And I'd like to use gyp.js by default some time later :smile: 

The switch should be the same for the `configure` and `build` commands, so env.variable seems preferable.

**Comment by indutny** [comment]
@pmed I would go with both, probably? Is it too hard to do it in an existing architecture?

Thanks for looking into it.

**Comment by pmed** [comment]
@indutny yes, both, with comman-line option and env var. It shouldn't be hard to keep branches both for gyp and gyp.js in configure and build commands.

I added the gyp.js dependency into node-gyp/package.json, and use gyp.main() on configure step. Build function spawns `ninja -C build/Release`  (or build/Debug). On Windows build function in node-gyp invokes vcvarsall.bat $target_arch before ninja.

I can try to make a PR into node-gyp.

**Comment by indutny** [comment]
Just run `npm test` on FreeBSD, everything appears to be working.

**Comment by refack** [comment]
@indutny what would you consider "others"? `sunos` and `aix-ppc`?
I can try to compile on a [raspberry-pi 2 running noobs](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bcm2836/README.md), and on a [MIPS32 OpenWRT machine](https://wiki.openwrt.org/doc/hardware/soc/soc.qualcomm.ar71xx)

**Comment by refack** [comment]
Well `llnode` was a wild goose chase, it doesn't compile cleanly on VS2015 with GYP...  :(
But I was able to compile `node.dll`, anybody know who do you test it?

**Comment by indutny** [comment]
@refack yeah solaris, aix, everywhere where node works.

**Comment by refack** [comment]
Cracking up my emulator...
